File: meris/harness/context.py
"""Context compression — message count + token budget."""

from __future__ import annotations

import json
from typing import Any

# Rough heuristic when tiktoken unavailable (~4 chars/token English/code mix)
CHARS_PER_TOKEN = 4


def estimate_tokens(text: str) -> int:
    if not text:
        return 0
    return max(1, len(text) // CHARS_PER_TOKEN)


def estimate_messages_tokens(messages: list[dict[str, Any]]) -> int:
    total = 0
    for m in messages:
        total += estimate_tokens(str(m.get("content") or ""))
        if m.get("tool_calls"):
            total += estimate_tokens(json.dumps(m["tool_calls"], ensure_ascii=False))
    return total


def truncate_content(text: str, max_tokens: int) -> str:
    budget_chars = max_tokens * CHARS_PER_TOKEN
    if len(text) <= budget_chars:
        return text
    keep = budget_chars - 40
    return text[:keep] + "\n...[meris: truncated for token budget]"


def shrink_tool_results(messages: list[dict[str, Any]], *, max_tool_tokens: int = 2000) -> list[dict[str, Any]]:
    """Truncate oversized tool role messages in-place copy."""
    out: list[dict[str, Any]] = []
    for m in messages:
        if m.get("role") == "tool":
            content = str(m.get("content") or "")
            if estimate_tokens(content) > max_tool_tokens:
                m = {**m, "content": truncate_content(content, max_tool_tokens)}
        out.append(m)
    return out


def compress_messages(
    messages: list[dict[str, Any]],
    *,
    max_messages: int = 48,
    max_tokens: int | None = None,
    max_tool_tokens: int = 2000,
) -> list[dict[str, Any]]:
    """Drop oldest turns; optionally enforce token budget."""
    msgs = shrink_tool_results(messages, max_tool_tokens=max_tool_tokens)

    if max_tokens is not None and estimate_messages_tokens(msgs) > max_tokens:
        msgs = _compress_by_tokens(msgs, max_tokens)

    if len(msgs) <= max_messages:
        return msgs

    system = [m for m in msgs if m.get("role") == "system"]
    rest = [m for m in msgs if m.get("role") != "system"]
    if not rest:
        return msgs

    first_user = 0
    for i, m in enumerate(rest):
        if m.get("role") == "user":
            first_user = i
            break

    head = rest[: first_user + 1]
    budget = max_messages - len(system) - len(head) - 1
    if budget < 4:
        budget = 4
    tail = rest[-budget:] if budget > 0 else []

    dropped = len(rest) - len(head) - len(tail)
    marker = {
        "role": "system",
        "content": f"[meris] Context compressed: dropped {dropped} older messages.",
    }
    return system + head + [marker] + tail


def _compress_by_tokens(messages: list[dict[str, Any]], max_tokens: int) -> list[dict[str, Any]]:
    """Keep system + first user task + newest messages within token budget."""
    system = [m for m in messages if m.get("role") == "system"]
    rest = [m for m in messages if m.get("role") != "system"]
    if not rest:
        return messages

    first_user_idx = 0
    for i, m in enumerate(rest):
        if m.get("role") == "user":
            first_user_idx = i
            break
    head = rest[: first_user_idx + 1]

    fixed = system + head
    fixed_tokens = estimate_messages_tokens(fixed)
    budget = max_tokens - fixed_tokens - estimate_tokens("[meris] token budget marker")
    if budget < 500:
        budget = 500

    tail: list[dict[str, Any]] = []
    used = 0
    for m in reversed(rest[first_user_idx + 1 :]):
        t = estimate_messages_tokens([m])
        if used + t > budget and tail:
            break
        tail.insert(0, m)
        used += t

    marker = {
        "role": "system",
        "content": (
            f"[meris] Token budget: kept ~{fixed_tokens + used} tokens "
            f"(limit {max_tokens}), dropped older turns."
        ),
    }
    return fixed + [marker] + tail

File: meris-rs/src/context.rs
//! Token estimation and message compression (parity with meris.harness.context).

use serde_json::{json, Value};

const CHARS_PER_TOKEN: usize = 4;

pub fn estimate_tokens(text: &str) -> usize {
    if text.is_empty() {
        return 0;
    }
    std::cmp::max(1, text.len() / CHARS_PER_TOKEN)
}

pub fn estimate_messages_tokens(messages: &[Value]) -> usize {
    let mut total = 0usize;
    for m in messages {
        let content = m.get("content").and_then(|c| c.as_str()).unwrap_or("");
        total += estimate_tokens(content);
        if let Some(tcs) = m.get("tool_calls") {
            if let Ok(s) = serde_json::to_string(tcs) {
                total += estimate_tokens(&s);
            }
        }
    }
    total
}

pub fn truncate_content(text: &str, max_tokens: usize) -> String {
    let budget_chars = max_tokens * CHARS_PER_TOKEN;
    if text.len() <= budget_chars {
        return text.to_string();
    }
    let keep = budget_chars.saturating_sub(40);
    let truncated: String = text.chars().take(keep).collect();
    format!("{truncated}\n...[meris: truncated for token budget]")
}

pub fn shrink_tool_results(messages: &[Value], max_tool_tokens: usize) -> Vec<Value> {
    messages
        .iter()
        .map(|m| {
            if m.get("role").and_then(|r| r.as_str()) == Some("tool") {
                let content = m.get("content").and_then(|c| c.as_str()).unwrap_or("");
                if estimate_tokens(content) > max_tool_tokens {
                    let mut copy = m.clone();
                    if let Some(obj) = copy.as_object_mut() {
                        obj.insert(
                            "content".into(),
                            json!(truncate_content(content, max_tool_tokens)),
                        );
                    }
                    return copy;
                }
            }
            m.clone()
        })
        .collect()
}

pub fn compress_messages(
    messages: &[Value],
    max_messages: usize,
    max_tokens: Option<usize>,
    max_tool_tokens: usize,
) -> Vec<Value> {
    let mut msgs = shrink_tool_results(messages, max_tool_tokens);

    if let Some(limit) = max_tokens {
        if estimate_messages_tokens(&msgs) > limit {
            msgs = compress_by_tokens(&msgs, limit);
        }
    }

    if msgs.len() <= max_messages {
        return sanitize_messages_for_api(&msgs);
    }

    let system: Vec<Value> = msgs
        .iter()
        .filter(|m| m.get("role").and_then(|r| r.as_str()) == Some("system"))
        .cloned()
        .collect();
    let rest: Vec<Value> = msgs
        .iter()
        .filter(|m| m.get("role").and_then(|r| r.as_str()) != Some("system"))
        .cloned()
        .collect();

    if rest.is_empty() {
        return msgs;
    }

    let first_user = rest
        .iter()
        .position(|m| m.get("role").and_then(|r| r.as_str()) == Some("user"))
        .unwrap_or(0);
    let head: Vec<Value> = rest[..=first_user].to_vec();
    let mut budget = max_messages
        .saturating_sub(system.len())
        .saturating_sub(head.len())
        .saturating_sub(1);
    if budget < 4 {
        budget = 4;
    }
    let tail_start = rest.len().saturating_sub(budget);
    let tail: Vec<Value> = if budget > 0 && tail_start > first_user + 1 {
        rest[tail_start..].to_vec()
    } else {
        vec![]
    };

    let dropped = rest.len().saturating_sub(head.len() + tail.len());
    let marker = json!({
        "role": "system",
        "content": format!("[meris] Context compressed: dropped {dropped} older messages."),
    });

    let mut out = system;
    out.extend(head);
    out.push(marker);
    out.extend(tail);
    sanitize_messages_for_api(&out)
}

fn drop_incomplete_tool_round(messages: &[Value]) -> Vec<Value> {
    let mut out: Vec<Value> = messages.to_vec();
    while out.last().and_then(|m| m.get("role")).and_then(|r| r.as_str()) == Some("tool") {
        out.pop();
    }
    if out.last().map(|m| {
        m.get("role").and_then(|r| r.as_str()) == Some("assistant")
            && m.get("tool_calls").is_some()
    }) == Some(true)
    {
        out.pop();
    }
    out
}

pub fn sanitize_messages_for_api(messages: &[Value]) -> Vec<Value> {
    let mut out: Vec<Value> = vec![];
    let mut pending: std::collections::HashSet<String> = std::collections::HashSet::new();

    for m in messages {
        if m.get("role").and_then(|r| r.as_str()) == Some("tool") {
            if let Some(tid) = m.get("tool_call_id").and_then(|t| t.as_str()) {
                if pending.remove(tid) {
                    out.push(m.clone());
                }
            }
            continue;
        }

        if !pending.is_empty() {
            out = drop_incomplete_tool_round(&out);
            pending.clear();
        }

        out.push(m.clone());

        if m.get("role").and_then(|r| r.as_str()) == Some("assistant") {
            pending.clear();
            if let Some(tcs) = m.get("tool_calls").and_then(|v| v.as_array()) {
                for tc in tcs {
                    if let Some(id) = tc.get("id").and_then(|v| v.as_str()) {
                        pending.insert(id.to_string());
                    }
                }
            }
        }
    }

    if !pending.is_empty() {
        out = drop_incomplete_tool_round(&out);
    }
    out
}

fn compress_by_tokens(messages: &[Value], max_tokens: usize) -> Vec<Value> {
    let system: Vec<Value> = messages
        .iter()
        .filter(|m| m.get("role").and_then(|r| r.as_str()) == Some("system"))
        .cloned()
        .collect();
    let rest: Vec<Value> = messages
        .iter()
        .filter(|m| m.get("role").and_then(|r| r.as_str()) != Some("system"))
        .cloned()
        .collect();

    if rest.is_empty() {
        return messages.to_vec();
    }

    let first_user_idx = rest
        .iter()
        .position(|m| m.get("role").and_then(|r| r.as_str()) == Some("user"))
        .unwrap_or(0);
    let head: Vec<Value> = rest[..=first_user_idx].to_vec();
    let mut fixed = system.clone();
    fixed.extend(head.clone());
    let fixed_tokens = estimate_messages_tokens(&fixed);
    let marker_tokens = estimate_tokens("[meris] token budget marker");
    let mut budget = max_tokens.saturating_sub(fixed_tokens + marker_tokens);
    if budget < 500 {
        budget = 500;
    }

    let mut tail: Vec<Value> = vec![];
    let mut used = 0usize;
    for m in rest[first_user_idx + 1..].iter().rev() {
        let t = estimate_messages_tokens(std::slice::from_ref(m));
        if used + t > budget && !tail.is_empty() {
            break;
        }
        tail.insert(0, m.clone());
        used += t;
    }

    let marker = json!({
        "role": "system",
        "content": format!(
            "[meris] Token budget: kept ~{} tokens (limit {max_tokens}), dropped older turns.",
            fixed_tokens + used
        ),
    });

    fixed.push(marker);
    fixed.extend(tail);
    sanitize_messages_for_api(&fixed)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn estimate_tokens_basic() {
        assert_eq!(estimate_tokens(""), 0);
        assert!(estimate_tokens("abcd") >= 1);
    }

    #[test]
    fn compress_by_token_budget() {
        let mut msgs = vec![json!({"role": "system", "content": "sys"})];
        msgs.push(json!({"role": "user", "content": "original task"}));
        for _ in 0..30 {
            msgs.push(json!({"role": "assistant", "content": "a".repeat(500)}));
            msgs.push(json!({"role": "user", "content": "u".repeat(500)}));
        }
        let out = compress_messages(&msgs, 100, Some(3000), 2000);
        assert!(estimate_messages_tokens(&out) <= 3500);
        assert!(out.iter().any(|m| {
            m.get("role").and_then(|r| r.as_str()) == Some("user")
                && m.get("content")
                    .and_then(|c| c.as_str())
                    .unwrap_or("")
                    .contains("original")
        }));
    }
}
